### ISSUE TITLE
`import/order` not respecting `builtin`

### ISSUE THREAD

**Issue by thernstig** [import-js/eslint-plugin-import#2687]
Using `"eslint-plugin-import": "2.27.5"`

```jsonc
// .eslintrc.json
{
  "root": true,
  "reportUnusedDisableDirectives": true,
  "parserOptions": {
    "ecmaVersion": 2023,
    "sourceType": "module"
  },
  "env": {
    "node": true,
    "es2022": true
  },
  "plugins": ["import"],
  "rules": {
    "import/order": [
      "error",
      {
        "groups": [
          [
            "builtin",
            "external",
            "internal",
            "parent",
            "sibling",
            "index",
            "object",
            "type"
          ]
        ]
      }
    ]
  }
}
```

```js
// test.js
import chpro from 'child_process';
import log4js from 'log4js';
import util from 'util';

const execFile = util.promisify(chpro.execFile);
const logger = log4js.getLogger('server');
```

I would expect `util` to be sorted before `log4js` but this is not the case. Neither is `node:util`.

**Comment by shawnmcknight** [comment]
I ran into this with a different built-in import (`path`).  If I changed the import to use the new node protocol semantics (i.e. `'node:path'`) then the import order was correct.  This made me believe that this detection might be due to a conflicting node core module and a module installed in `node_modules`.  Sure enough, if I look at my `node_modules` there is a `path` dependency in there (also a `util` dependency, which might match @thernstig 's scenario).

If I use a core module like `fs` where I don't have a conflicting dependency in `node_modules` then the import order is correct (actually wanted to precede `node:path` which is what I would have expected).

Taking a guess here that `import/order` might be getting confused because of the duplicate names in core and `node_modules`, but that's just a shot in the dark guess.

**Comment by thernstig** [comment]
@shawnmcknight that is not my experience. The above config does not complain about this ordering:

```js
import express from 'express';
import log4js from 'log4js';
import chpro from 'node:child_process';
import fsp from 'node:fs/promises';
```

I would expect all `node:` to end up first.

**Comment by ljharb** [comment]
@thernstig why do you have the nested array in `groups`?

**Comment by ljharb** [comment]
I was able to partially fix this in 1fa29717518732e3143aaf8155713133c383bf3c - now the same error shows up with a nested group or a flat one - but the ordering still isn't fully correct.

**Comment by thernstig** [comment]
@ljharb I stole that from eslint-config-airbnb-base, so maybe you can answer that. I saw though that https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/order.md does not have the nested array.

But I am using eslint-config-airbnb-base so in the end I get my config from there.

**Comment by ljharb** [comment]
@thernstig the base config has `{ groups: [['builtin', 'external', 'internal']] }` in it, so something else must be providing that list.

**Comment by thernstig** [comment]
@ljharb I took it from eslint-config-airbnb-base, but then just added all items from https://github.com/import-js/eslint-plugin-import/blob/683d3a5927adf97a1654cfb4d9d3caf148e1eb8a/docs/rules/order.md?plain=1#L105. Never reflected about the array since eslint-config-airbnb-base had the extra array.

**Comment by ljharb** [comment]
No, it’s not wrong, i was just curious where the config comes from.

**Comment by yndajas** [comment]
👋🏻 is the (partial) fix for this correct, particularly [this conditional](https://github.com/import-js/eslint-plugin-import/commit/1fa29717518732e3143aaf8155713133c383bf3c#diff-2825fcc427c846fe18d501930e1a3d83be722027f1f643801afdcc2c255dd4a8R430-R433)?

After updating to 2.28.0 we're seeing `"groups": [["builtin", "external"]]` (one group) lead to builtin and external being separated, when I believe these should be grouped together with everything else in a second group. This was the behaviour before 2.28.0 and seemingly [what the docs suggest](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/order.md#groups-array): `['sibling', 'parent'], // Then sibling and parent types. They can be mingled together` ("They can be mingled together"). We've had to [amend the config](https://github.com/ministryofjustice/hmpps-accredited-programmes-ui/pull/129/commits/ab5a73e73b4a0a0223ddf0eb3c2362db81cc12f2) to keep the previous behaviour

I might not be fully understanding the original issue in this thread, but I would've thought it would be solved by de-nesting? It looks like the [Airbnb config example](https://github.com/airbnb/javascript/blob/cda44dad403979d313a4b02ce55f0bb6cc3a24b2/packages/eslint-config-airbnb-base/rules/imports.js#L149) is: `groups: [['builtin', 'external', 'internal']]`, which would group builtin, external, and internal and then follow with everything else in a second group, so if you wanted builtin then external then internal then parent then sibling then index then object then type it would be `groups: ["builtin", "external", "internal", "parent", "sibling", "index", "object", "type"]` (de-nested)

I note that there's a TODO in the conditional and this is a described as a partial fix, so maybe the new behaviour is intended to be reverted eventually?

**Comment by ljharb** [comment]
@yndajas that sounds right to me. If the behavior has changed in v2.28.0 then that's a bug that needs fixing.

**Comment by yndajas** [comment]
@ljharb glad I'm not confused! Shall I open a new issue or leave it with you?

**Comment by ljharb** [comment]
@yndajas I think this issue probably covers it; a PR would be most helpful :-)

**Comment by thernstig** [comment]
@yndajas @ljharb I am not sure how the problem reported there solves this issue? This is about it not respecting `builtin` at all.

**Comment by ljharb** [comment]
@thernstig when it's nested like that, all of those groups can be intermingled. The config needs to de-nest the groups in order for their order to be enforced.

**Comment by yndajas** [comment]
@thernstig per @ljharb, I believe the original issue in this thread would be solved by denesting the `groups`:

```
// .eslintrc.json
{
  "root": true,
  "reportUnusedDisableDirectives": true,
  "parserOptions": {
    "ecmaVersion": 2023,
    "sourceType": "module"
  },
  "env": {
    "node": true,
    "es2022": true
  },
  "plugins": ["import"],
  "rules": {
    "import/order": [
      "error",
      {
        "groups": [
          "builtin",
           "external",
          "internal",
          "parent",
          "sibling",
          "index",
          "object",
          "type"
        ]
      }
    ]
  }
}
```

More detailed explanation in [my earlier comment](https://github.com/import-js/eslint-plugin-import/issues/2687#issuecomment-1671038558)

Hope that clarifies!

**Comment by ljharb** [comment]
Per that comment, I'll close this, but can reopen if it doesn't work out.
